### ISSUE TITLE
Liquibase 4.26.0 multi-threading issue caused by cached prepared statements

### ISSUE THREAD

**Issue by adaryin** [liquibase/liquibase#5635]
### Search first

- [X] I searched and no similar issues were found

### Description

**Problem**

Liquibase 4.26.0 cannot be safely used for loading data in a multi-threaded environment working either with a single or multiple databases.
First I've applied the solution from [here](https://forum.liquibase.org/t/concurrent-updating-of-databases-is-no-longer-possible-since-upgrading-to-liquibase-4-x/7013), however the problem lies in a bad design of caching prepared statements in `ExecutablePreparedStatementBase` class. 
Our framework  uses `LoadDataChange` feature to populate data to our db's from CSV files in multiple threads. After upgrade to Liquibase 4.26.0 from 3.8.9 our solution can load data to db's in a single thread only. 

**Root cause in class**
```
public abstract class ExecutablePreparedStatementBase implements ExecutablePreparedStatement {
    ...
    //Some databases do extra work on creating prepared statements, so constantly creating new prepared statements is expensive
    //When running through a CSV file, the SQL will be the same within the same file so just storing the last seen prepared statement is all we need
    //Ideally the creation of the prepared statements would happen at the spot where we know we should be re-using it and that code can close it.
    // But that will have to wait for a refactoring of this code.
    // So for now we're trading leaving at most one prepared statement unclosed at the end of the liquibase run for better re-using statements to avoid overhead
    private static PreparedStatement lastPreparedStatement;
    private static String lastPreparedStatementSql;
   ...
    protected PreparedStatement getCachedStatement(String sql) {
        if (lastPreparedStatement == null || lastPreparedStatementSql == null) {
            return null;
        }

        boolean statementIsValid = true;
        if (lastPreparedStatementSql.equals(sql)) {
        ...
        } else {
            statementIsValid = false;
        }

        if (!statementIsValid) {
            JdbcUtil.closeStatement(lastPreparedStatement);
            lastPreparedStatement = null;
            lastPreparedStatementSql = null;
        }

        return lastPreparedStatement;
    }
```
The fields `lastPreparedStatement` / `lastPreparedStatementSql` are declared as static and therefore they affect all descendants of `ExecutablePreparedStatementBase` no matter of object clones. The worst thing is closing previously cached statements on every SQL query string change. 
When used concurrently against multiple databases SQL in statements will be different in 99.99% and this causes unexpected issues like: an attempt to use already closed statements, various NPE's because some resources get freed inside a statement, etc...
In our environment we update MSSQL, DB2 AS/400, Oracle db's in parallel using different Liquibase objects and the above issue caused the used statements to be randomly closed so I got plenty of different errors with the same root cause: `JdbcUtil.closeStatement(lastPreparedStatement)`.

Please consider use of thread local variables and get rid of static fields as much as possible. I am willing to contribute if you consider this possible. 

**Workaround**

1. Create a custom descendant to LoadDataChange
```
@DatabaseChange(name = "loadData", description = "Disables use of cached prepared statements",
        priority = 100, appliesTo = "table"
)
public class NoCacheLoadDataChange extends LoadDataChange {

    @Override
    protected ExecutablePreparedStatementBase createPreparedStatement(Database database, String catalogName, String schemaName, String tableName, List<LoadDataColumnConfig> columns, ChangeSet changeSet, ResourceAccessor resourceAccessor) {
        return new NoCacheInsertPrepareStatement(database, catalogName, schemaName, tableName, columns, changeSet, resourceAccessor);
    }

    public static class NoCacheInsertPrepareStatement extends InsertExecutablePreparedStatement {
    ...

        @Override
        protected PreparedStatement getCachedStatement(String sql) {
            return null;
        }
    }

    @Override
    protected SqlStatement[] generateStatementsFromRows(Database database, List<LoadDataRowConfig> rows) {
    ...
                } else {
                    return new SqlStatement[]{
                            new NoCacheBatchDmlExecutablePreparedStatement(
                                    database, getCatalogName(), getSchemaName(),
                                    getTableName(), getColumns(),
                                    getChangeSet(), Scope.getCurrentScope().getResourceAccessor(),
                                    preparedStatements)
                    };
                }
            } else {
                return statements.toArray(SqlStatement.EMPTY_SQL_STATEMENT);
            }
    ...
    }

    private static class NoCacheBatchDmlExecutablePreparedStatement extends BatchDmlExecutablePreparedStatement {
    ...
        @Override
        protected PreparedStatement getCachedStatement(String sql) {
            return null;
        }
    }
}
```
2. Register it under META-INF/services

`BatchDmlExecutablePreparedStatement` and `InsertExecutablePreparedStatement` classes don't have a common ancestor and this resulted in overriding of `getCachedStatement` method twice. 

### Steps To Reproduce

No formal steps because of a multi-threading nature of the problem. 
A few moments. 
1. All our Liquibase instances used for loading data into multiple databases are isolated
2. All instances use independent changelogs that look like:
```
...
  - changeSet:
      id: my_id
      author: me
      changes:
        - loadData:
            encoding: enc
            file: csv/FILE.csv
...
```
3. Data load is conducted in parallel

### Expected/Desired Behavior

Caching for prepared statements is conducted independently per a thread. 

### Liquibase Version

4.26.0

### Database Vendor & Version

N/A

### Liquibase Integration

gradle

### Liquibase Extensions

N/A

### OS and/or Infrastructure Type/Provider

N/A

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

**Comment by TedKeenan** [comment]
I would like to up vote that it is a really bad idea to cache the connection in this way. At a minimum the variables should be class local so that when an instance of the class goes out of scope the prepared statement goes with them. 
I find it hard to believe that the performance justifies the side issues that this creates.

**Comment by filipelautert** [comment]
Hello @adaryin , thanks for the detailed issue. I like the idea of using ThreadLocal instead of static only for    lastPreparedStatement,, lastPreparedStatementSql and executeWithFlagsMethod - with 4.27.0 we started to use it in other locations too such as Scope management to improve our multithreading support.
Would you be able to submit a PR with this change? Then just link here and tag me to get it moving. Thanks!

**Comment by adaryin** [comment]
Hi @filipelautert, thanks for the reply. I would like to submit a PR (it's ready), but on `git push` I get:
Permission to liquibase/liquibase.git denied to adaryin
Could someone pls assist?

**Comment by filipelautert** [comment]
> Hi @filipelautert, thanks for the reply. I would like to submit a PR (it's ready), but on `git push` I get: `Permission to liquibase/liquibase.git denied to adaryin` Could someone pls assist with granting rights to my account?

Hello @adaryin - in fact you need to `fork` the project on main page ( https://github.com/liquibase/liquibase/fork ), work on your fork and then open the PR . When you open the PR it will automatically point to `liquibase/liquibase` project 👍

**Comment by adaryin** [comment]
Filipe, thanks for pointing this out. Will do as suggested.

**Comment by adaryin** [comment]
@filipelautert here you go:
https://github.com/liquibase/liquibase/pull/5874
